src/Database/Driver.php

```diff
@@ -215,6 +225,73 @@ public function exec(string $sql): int|false
      */
     abstract public function enabled(): bool;
 
+    /**
+     * @inheritDoc
+     */
+    public function execute(string $sql, array $params = [], array $types = []): StatementInterface
+    {
+        $statement = $this->prepare($sql);
+        if (!empty($params)) {
+            $statement->bind($params, $types);
+        }
+        $this->executeStatement($statement);
+
+        return $statement;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function run(Query $query): StatementInterface
+    {
+        $statement = $this->prepare($query);
+        $query->getValueBinder()->attachTo($statement);
+        $this->executeStatement($statement);
+
+        return $statement;
+    }
+
+    /**
+     * Execute the statement and log the query string.
+     *
+     * @param \Cake\Database\StatementInterface $statement Statement to execute.
+     * @return void
+     */
+    protected function executeStatement(StatementInterface $statement): void
+    {
+        if ($this->logger === null) {
+            $statement->execute();
+
+            return;
+        }
+
+        $exception = null;
+        $took = 0.0;
+
+        try {
+            $start = microtime(true);
+            $statement->execute();
+            $took = (float)number_format((microtime(true) - $start) * 1000, 1);
+        } catch (PDOException $e) {
+            $exception = $e;
+        }
+
+        $logContext = [
+            'driver' => $this,
+            'error' => $exception,
+            'params' => $statement->getBoundParams(),
+        ];
+        if (!$exception) {
+            $logContext['numRows'] = $statement->rowCount();
+            $logContext['took'] = $took;
+        }
+        $this->logQuery($statement->queryString(), $logContext);
+
+        if ($exception) {
+            throw $exception;
+        }
```

Comment:
Should this be a new protected method? Having a protected method gives application/plugin developers a simple 'slot' to implement if they want to replace how a statement is logged.

Reply:
Well they could just override `executeStatement()`, it's only couple of extra lines at the top to check if logging is enabled. Earlier I had the check in the caller methods but othercorey suggested moving it inside this method itself.

Reply:
I don't think it would be a common extension point either. We could add the protected hook method if/when we get feedback from plugin authors that need to override the behavior.